src/MySqlConnector/MySqlCommand.cs

```diff
@@ -422,6 +427,58 @@ private void CancelCommandForTimeout()
 		Connection?.Cancel(this, m_commandId, false);
 	}
 
+	public void ResetCommandTimeout()
+	{
+		if (m_reset == null)
+			ResetCommandTimeoutDefault();
+		else m_reset();
+	}
+
+	/// <summary>
+	/// Causes the effective command timeout to be reset back to the value specified by <see cref="ICancellableCommand.CommandTimeout"/>
+	/// plus <see cref="MySqlConnectionStringBuilder.CancellationTimeout"/>. This allows for the command to time out, a cancellation to attempt
+	/// to happen, then the "hard" timeout to occur.
+	/// </summary>
+	/// <remarks>As per the <a href="https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlcommand.commandtimeout.aspx">MSDN documentation</a>,
+	/// "This property is the cumulative time-out (for all network packets that are read during the invocation of a method) for all network reads during command
+	/// execution or processing of the results. A time-out can still occur after the first row is returned, and does not include user processing time, only network
+	/// read time. For example, with a 30 second time out, if Read requires two network packets, then it has 30 seconds to read both network packets. If you call
+	/// Read again, it will have another 30 seconds to read any data that it requires."
+	/// The <see cref="ResetCommandTimeout"/> method is called by public ADO.NET API methods to reset the effective time remaining at the beginning of a new
+	/// method call.</remarks>
+	private void ResetCommandTimeoutDefault()
+	{
+		var session = Connection?.Session;
+		if (session is not null)
+		{
+			if (CommandTimeout == 0)
+			{
+				session.SetTimeout(Constants.InfiniteTimeout);
+				m_reset = m_noOp;
+			}
+			else
+			{
+				((ICancellableCommand) this).SetTimeout(CommandTimeout * 1000);
+				session.SetTimeout((session.CancellationTimeout <= 0 ?
+					                   CommandTimeout : CommandTimeout + session.CancellationTimeout)
+				                   * 1000);
+				m_reset = ResetCommandTimeoutSet;
+			}
+		}
+	}
+
+	private void ResetCommandTimeoutSet()
```

Comment:
Method names should generally take the form `VerbNoun`. I'm not sure what `ResetCommandDefaultSet` means, exactly.

Reply:
this may be not clear on first read, but this "ResetCommandDefaultSet" is the first "ResetCommandTimeout" that is called, setting m_reset for further  call of ResetCommandTimeout (if CommandTimeout is not changed)